object-construction-checker/tests/socket/InputOutputStreams.java

```diff
@@ -0,0 +1,100 @@
+// A MustCallChoice test wrt sockets. Also coincidentally tests that MCC sets can be larger than two.
+
+import java.io.*;
+import java.net.*;
+import org.checkerframework.checker.objectconstruction.qual.*;
+import java.io.IOException;
+
+class InputOutputStreams {
+    void test_close_sock(@Owning Socket sock) throws IOException {
+        try{
+            InputStream is = sock.getInputStream();
+            OutputStream os = sock.getOutputStream();
+        } catch (IOException e) {
+
+        } finally {
+            sock.close();
+        }
+    }
+
+    // :: error: required.method.not.called
+    void test_close_is(@Owning Socket sock) throws IOException {
+        InputStream is = null;
+        OutputStream os = null;
+        try {
+            is = sock.getInputStream();
+            os = sock.getOutputStream();
+        } catch (IOException e) {
+
+        } finally {
+            is.close();
+        }
+    }
+
+    // :: error: required.method.not.called
+    void test_close_os(@Owning Socket sock) throws IOException {
+        InputStream is = sock.getInputStream();
+        OutputStream os = sock.getOutputStream();
+        os.close();
+    }
+
+    // :: error: required.method.not.called
+    void test_close_os2(@Owning Socket sock) throws IOException {
+        OutputStream os = null;
+        InputStream is = null;
+        try {
+            is = sock.getInputStream();
+        } catch (IOException e) {
+            try {
+            os = sock.getOutputStream();
+            } catch (IOException ee) {}
+        } finally {
+            os.close();
+        }
+    }
+
+    // :: error: required.method.not.called
+    void test_close_os3(@Owning Socket sock) throws IOException {
```

Comment:
Can we add this version of this test:

```
void test_close_os4(@Owning Socket sock) throws IOException {
        OutputStream os = null;
        try {
            InputStream is = sock.getInputStream();
        } catch (IOException e) { }
        try {
            os = sock.getOutputStream();
        } finally {
            if (os != null) {
                os.close();
            } else {
                sock.close();
            }
        }
    }
```

I think this version should pass.

Reply:
It's a great example! To pass this case, we should handle the boolean condition problem.

Reply:
To further clarify, in the case where `getOutputStream()` throws an `IOException`, `os` does _not_ get added to the set of "choice" variables for the socket resource.  Further, the `os` variable remains definitely `null`.  When our worklist analysis gets to the `finally` block along the corresponding CFG edge, it does not know that `os` is definitely null, and it is only tracking `sock` as a name for the resource.  So, it analyzes a path through the "then" branch of the conditional, where `sock.close()` is not invoked, and reports an error.  Addressing this issue requires more general tracking of additional boolean conditions (in this case, regarding the nullness of `os`), which we should leave outside the scope of this pull request.

Reply:
Thanks for the explanation. If possible, I'd like to keep the test case with a note about why it fails (probably exactly what you wrote in the comment above) - I suspect it will be useful later.